analyzers/src/SonarAnalyzer.Common/Helpers/ArgumentDescriptor.cs

```diff
@@ -0,0 +1,177 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Helpers;
+
+public enum InvokedMemberKind
+{
+    Method,
+    Constructor,
+    Indexer,
+    Attribute
+}
+
+public class ArgumentDescriptor
+{
+    private ArgumentDescriptor(InvokedMemberKind memberKind, Func<IMethodSymbol, bool> invokedMemberConstraint, Func<string, StringComparison, bool> invokedMemberNameConstraint,
+        Func<SemanticModel, ILanguageFacade, SyntaxNode, bool> invokedMemberNodeConstraint, Func<IReadOnlyCollection<SyntaxNode>, int?, bool> argumentListConstraint,
+        Func<IParameterSymbol, bool> parameterConstraint, RefKind? refKind)
+    {
+        MemberKind = memberKind;
+        ArgumentListConstraint = argumentListConstraint;
+        RefKind = refKind;
+        ParameterConstraint = parameterConstraint;
+        InvokedMemberNameConstraint = invokedMemberNameConstraint;
+        InvokedMemberNodeConstraint = invokedMemberNodeConstraint;
+        InvokedMemberConstraint = invokedMemberConstraint;
+    }
+
+    public static ArgumentDescriptor MethodInvocation(KnownType invokedType, string methodName, string parameterName, int argumentPosition) =>
```

Comment:
Definitly. I started working on one, but the builder is quite complicated (it has quite a complicated inheritance hierarchy because of the different member-kinds and because some settings are mandatory while others are not). At the end of the day, the matching logic is the important part of the PR. The builder could be another effort.